Add skip_first_wait to profiler.schedule

sraikund16 opened the pull request:
Summary:
Addresses https://github.com/pytorch/pytorch/issues/91888

We use wait as the amount you wait in between cycles when profiling and skip_first to delay the start of said profiling. However, once skip_first steps are completed, we immediately go to the wait phase. This is not problematic if wait is smaller than skip_first because we can just lower the values of skip_first, but if it is larger then we end up starting the first profile much later than desired. For example imagine a skip first of 1 and a wait of 100 with repeat of 2. We do want to wait 100 steps in between cycle 1 and 2 but we may not want to start warmup of cycle 1 at step 101 (forced because wait occurs directly after first steps skipped). This diff addresses this by adding a flag to skip the first wait.

Adds new flag but sets to false by default so that existing impl is not affected.

Test Plan:
Got reasonable traces with this schedule:


schedule=torch.profiler.schedule(
            wait=10, warmup=3, active=1, repeat=1, skip_first=1, skip_first_wait=1
        )

D66198138

@diff-train-skip-merge

aaronenyeshi commented on torch/profiler/profiler.py:
```diff
@@ -434,14 +434,22 @@ class ProfilerAction(Enum):
     The optional number of cycles is specified with the ``repeat`` parameter, the zero value means that
-    the cycles will continue until the profiling is finished.
+    the cycles will continue until the profiling is finished. The ``skip_first_wait`` parameter controls
+    whether the first wait stage should be skipped. This can be useful if a user wants to wait larger than skip_first
```
may be typo: larger == longer

sraikund16 replied:
grammar issue will fix

aaronenyeshi commented on torch/profiler/profiler.py:
```diff
@@ -434,14 +434,22 @@ class ProfilerAction(Enum):
-    the cycles will continue until the profiling is finished.
+    the cycles will continue until the profiling is finished. The ``skip_first_wait`` parameter controls
+    whether the first wait stage should be skipped. This can be useful if a user wants to wait larger than skip_first
+    between cycles but not for the first profile.
```
would be helpful to have some example here for the wait and skip_first occurances

sraikund16 replied:
added example in newest diff

briancoutinho reviewed (approved):
lgtm :)

sraikund16 committed:
Add skip_first_wait to profiler.schedule (#141070)

Summary:

Addresses https://github.com/pytorch/pytorch/issues/91888

We use `wait` as the amount you wait in between cycles when profiling and `skip_first` to delay the start of said profiling. However, once `skip_first` steps are completed, we immediately go to the `wait` phase. This is not problematic if `wait` is smaller than `skip_first` because we can just lower the values of `skip_first`, but if it is larger then we end up starting the first profile much later than desired. For example imagine a skip first of 1 and a `wait` of 100 with repeat of 2. We do want to wait 100 steps in between cycle 1 and 2 but we may not want to start warmup of cycle 1 at step 101 (forced because `wait` occurs directly after first steps skipped). This diff addresses this by adding a flag to skip the first wait.

Adds new flag but sets to false by default so that existing impl is not affected.

Test Plan:
Got following traces with this schedule:


schedule=torch.profiler.schedule(
            wait=10, warmup=3, active=1, repeat=1, skip_first=1, skip_first_wait=1
        )

https://www.internalfb.com/intern/perfdoctor/trace_view?filepath=tree/traces/dynocli/devvm2185.cco0.facebook.com/rank-0.Nov_19_14_38_09.417495.pt.trace.json.gz&bucket=gpu_traces

https://www.internalfb.com/intern/perfdoctor/trace_view?filepath=tree/traces/dynocli/devvm2185.cco0.facebook.com/rank-0.Nov_19_14_29_32.337974.pt.trace.json.gz&bucket=gpu_traces

Reviewed By: aaronenyeshi, briancoutinho

Differential Revision: D66198138

izaitsevfb commented:
@pytorchbot merge -i

izaitsevfb commented:
@pytorchbot merge -f 'landed internally'

izaitsevfb commented:
@pytorchbot revert -m "oops, it's actually reverted internally" -c ghfirst
